DESIGN.md

```diff
@@ -0,0 +1,756 @@
+internally. It is possible to have two operations in-flight from a single
+resource associated with different queues and threads.
+
+Because operations are not `Send`, tasks awaiting these operations are also not
+`Send`, making them unable to be spawned using the spawn function from the Tokio
+crate. The tokio-uring crate will provide a spawn function that accepts not
+`Send` tasks.
+
+When using multiple isolated runtime threads, balancing load between them
+becomes challenging. Applications must take care to ensure load remains balanced
+across threads, and strategies tend to vary. For example, a TCP server can
+distribute accepted connections across multiple threads, ideally while
+maintaining equal load across threads. One approach is to submit "accept"
+operations for the same listener concurrently across all the runtime threads
+while ensuring overloaded workers back-off. Defining runtime load is out of this
+crate's scope and left to the application, though pseudocode follows.
+
+```rust
+let listener = Arc::new(TcpListener::bind(...));
+
+spawn_on_each_thread(async {
+    loop {
+        // Wait for this worker to have capacity. This
+        // ensures there are a minimum number of workers
+        // in the runtime that are flagged as with capacity
+        // to avoid total starvation.
+        current_worker::wait_for_capacity().await;
+        
+        let socket = listener.accept().await;
+        
+        spawn(async move { ... });
+    }
+})
+```
+
+## Reading and writing
+
+Read and write operations require passing ownership of buffers to the kernel.
+When the operation completes, the kernel returns ownership of the buffer to the
+caller. The caller is responsible for allocating the buffer's memory and
+ensuring it remains alive until the operation completes. Additionally, while the
+kernel owns the memory, the process may not read from or write to the buffer. By
+designing the Rust APIs using ownership passing, Rust enforces the requirements
+at compile time. The following example demonstrates reading and writing with a
+file resource.
+
+```rust
+use tokio_uring::buf;
+
+/// The result of an operation that includes a buffer. The buffer must
+/// be returned to the caller when the operation completes successfully
+/// or fails.
+///
+/// This is implemented as a new type to implement std::ops::Try once
+/// the trait is stabilized.
+type BufResult<T> = (std::io::Result<T>, buf::Slice);
+
+// Positional read and write function definitions
+impl File {
+    async fn read_at(&self, buf: buf::Slice, pos: u64) -> BufResult<usize>;
+    
+    async fn write_at(&self, buf: buf::Slice, pos: u64) -> BufResult<usize>;
+}
+
+/// The caller allocates a buffer
+let buf = buf::IoBuf::with_capacity(4096);
+
+// Read the first 1024 bytes of the file, when `std::ops::Try` is
+// stable, the `?` can be applied directly on the `BufResult` type.
+let BufResult(res, buf) = file.read_at(0, buf.slice(0..1024)).await;
```

Comment:
Is there any reason why we don't use `Index` here rather than having a separate slice function? We could do like `buf[0..1024]` and both should panic when the start or end index is out of the capacity.

Reply:
It's probably because `Index` is required to return a reference but `.slice` has to return an owned value.